### ISSUE TITLE
A simpler polysemy fork

### ISSUE THREAD

**Issue by FongHou** [polysemy-research/polysemy#229]
Hello, polysemy community. Awesome library!

I have created a polysemy fork [here](https://github.com/FongHou/freer-simple).  It's primarily for retaining a simple and fast (prior to ghc 8.10) copy of polysemy freer monad effect encoding without the high-order complexity (@isovector did a great job to provide ergonomics, but implementation is still too advanced for me :-). I'm planning to continue improving the robustness and performance, adding more tests and benchmarks. 

I'd really appreciate any feedback coming from experts here!

**Comment by isovector** [comment]
Hi there! Thanks for your work backporting this stuff to `freer-simple`, it's looking great!

That being said, I'm wondering what your goals are here; maybe we can accomplish them without needing to fork.

**Comment by FongHou** [comment]
Hi @isovector,

The goal is for people like me to try out algebraic effects based design in larger real work apps. Without fear of hitting walls, then getting completely stuck. Because of two reasons:

1. Polysemy's high-order effects require some very advanced ghc knowledge. To be honest, I still have no clue why/how those Tatictic/Strategies work, and when/where those Embed/Final combinators should or should not be used.  

2. Until ghc 8.10 (with your patch) release,  polysemy performance seems around three orders of magnitude slower than mtl/fused-effects (see [effects-benchmarks](https://github.com/patrickt/effects-benchmarks). If some performance issues occurred in a large codebase in ghc < 8.10, in addition to looking into data structures/algos and laziness space leaks, the usual suspects, it'll be a burden to worry about that polysemy **might** be another root cause. 

In summary, this backport puts approachability, simplicity (w.r.t debugging type errors and performance bugs) above all other goals (articulated nicely by Patrick Thomson https://www.youtube.com/watch?v=vfDazZfxlNs).

Thank you very much for listening and considering. It'll be great a sister package like polysemy-simple exist under polysemy-research. With that said, I don't mean to distract your time and other goals you have on polysemy. I'll be happy to keep this backport. Please let me know if I'm doing anything wrong there.

Thanks!

**Comment by isovector** [comment]
If you restrict yourself to first order effects --- which is what your fork does AFAICT --- you can completely ignore strategies and tactics. As such, it seems to me like you're tying your arms behind your back for no reason on this front!

The performance is more of a compelling reason (though I'd be worried if anyone who desperately needs this performance is feckless enough to use software as experimental as polysemy.) I'd suggest a better approach is to convince @lexi-lambda to merge https://github.com/lexi-lambda/freer-simple/pull/24 into `freer-simple`; such a thing will give you MTL-order performance in an already mature effects library.

Moving forwards, I'd recommend we take one or more of the following directions:

1. If newcomers really do have such a fear of the higher-order effects, let's introduce a new module `Polysemy.Simple` that only exposes first-order combinators. Then people can ignore the harder pieces of the library, without fundamentally closing the door on those capabilities.

2. Get https://github.com/lexi-lambda/freer-simple/pull/24 into a good enough shape that Lexi will merge it, and then switch to using `freer-simple`. The effects ecosystem is already fragmented enough that I'm very hesitant to endorse splitting it further. From here, I suspect it would be relatively easy to share interpretations between `polysemy` and `freer-simple`.

3. If neither of the above are satisfying to you, I wish you all the best of luck in your fork! Though I'd ask you not use the `polysemy` branding, since the two projects have nothing to do with one another.

Cheers!

**Comment by FongHou** [comment]
Thank you for your advice, @isovector! 

The back-porting actually started with merging your too-fast-too-free pull request into a freer-simple fork. The only incompatible change I made is swapping State (a, s) tuple to (s, a). And changed a few functions, intercept and interceptS, back to freer-simple's naming (interpose, InterposeState). So the fork actually is a MTL-order performance freer-simple!

Regarding your recommendations:

1. If Polysemy.Simple only exposes first-order combinators, but still uses higher-order core encoding extended after your too-fast-too-simple, it wouldn't get back MTL-order performance, correct? If that's the case, it seems not worth the effort. 

2. Agree. I'll try to get my fork into a good enough shape. BTW, the interaction between bracket (ResourceT based) and State/Writer is the area I don't feel comfortable with. For example, this [line] (https://github.com/FongHou/freer-simple/blob/e0327d0c4a3de16a5bd1ff379bb5a392b59eacfe/specs/BracketSpec.hs#L31) won't compile (error below). And I can't figure out how to make LiftBracket/runBracket vary return types in order to thread (local get/put, listen/censor) states through chain of interpreters.

```
    • Occurs check: cannot construct the infinite type:
        x ~ ([String], x)
      Expected type: Control.Monad.Freer.Internal.Eff
                       '[Input Int, Input [Char], Output String, Output String,
                         Error Bool, Error String, Trace, IO]
                       x
                     -> Control.Monad.Freer.Internal.Eff '[IO] x
        Actual type: Control.Monad.Freer.Internal.Eff
                       '[Input Int, Input [Char], Output String, Output String,
                         Error Bool, Error String, Trace, IO]
                       x
                     -> Control.Monad.Freer.Internal.Eff '[IO] ([String], x)
```

3. I'll change the branch name to something like freer-simple-faster.

Cheers!

**Comment by TheMatten** [comment]
@FongHou Hmm, I still don't understand what makes fork without higher-order effects useful - you can just ignore `Tactics` stuff in `polysemy`, while still being able to use builtin or contributed higher-order effects.

**Comment by FongHou** [comment]
Hi @TheMatten, the fork is mostly about getting back MTL-performance, temporarily. And as a learning step for me to eventually enjoy the full polysemy capabilities. 

Cheers!
